### ISSUE TITLE
NgClass removes existing class when applying array that contains with 2 empty slots

### ISSUE THREAD

**Issue by wizarrc** [angular/angular#17852]
## I'm submitting a ...

<pre><code>
[ ] Regression (behavior that used to work and stopped working in a new release)
[x] Bug report 
[ ] Feature request
[ ] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
</code></pre>

**Construct a class array with 4 strings. The middle classes are an empty string. Then change the 2nd class, leaving the other classes unchanged.**

## Current behavior

NgClass does not apply correct classes. It removes the last class from the array. Internally the last class is added to the add and remove list, and since the add class code does nothing since it's already in the list and does not error, the remove list removes the class.

## Expected behavior

NgClass apply correct classes. The action should be to do nothing for the other existing classes and only add the 2nd class.

The following code tries to add the existing class in `changes.forEachAddedItem`, then removes it in `changes.forEachRemovedItem`. I would suggest not only fixing the bug, but reverse the order to first remove, then add, in case there are more bugs in the future.

https://github.com/angular/angular/blob/master/packages/common/src/directives/ng_class.ts#L105
```
private _applyIterableChanges(changes: IterableChanges<string>): void {
    changes.forEachAddedItem((record) => {
      if (typeof record.item === 'string') {
        this._toggleClass(record.item, true);
      } else {
        throw new Error(
            `NgClass can only toggle CSS classes expressed as strings, got ${stringify(record.item)}`);
      }
    });

    changes.forEachRemovedItem((record) => this._toggleClass(record.item, false));
}
```

## Minimal reproduction of the problem with instructions

http://plnkr.co/edit/bTsnyx1EciH66tRROu5o?p=preview

## What is the motivation / use case for changing the behavior?

Developers expect binding to be correct.

## Please tell us about your environment

<pre><code>
Angular version: 4.2.4

Browser:
- [x] Chrome (desktop) version 59
- [ ] Chrome (Android) version XX
- [ ] Chrome (iOS) version XX
- [ ] Firefox version XX
- [ ] Safari (desktop) version XX
- [ ] Safari (iOS) version XX
- [ ] IE version XX
- [x] Edge version 15
 
For Tooling issues:
- Node version: XX 
- Platform: 

Others:

</code></pre>

**Updated all NgStyle to NgClass** - sorry for the brain fart.

**Comment by tytskyi** [comment]
1. Incorrect issue title, you do refer to `[ngStyle]`, not `NgStyle `

2. > Developers expect binding to be correct.

    To not have such surprises do not use `[ngStyle]`

    see https://github.com/angular/angular/issues/13940#issuecomment-276079860

    and https://github.com/angular/angular/pull/13958#issuecomment-276245722

Perfect use-case for `[ngStyle]` - is to generate dynamic class-name in runtime. When you don't know class in advance, for example `[ngClass]="getMyRandomClassName()"`. Or attaching multiple classes at once. Everything else could be solved within `[class.class-name]="true"` binding.

Related issues that i found quickly: 
- https://github.com/angular/angular/issues/15396
- https://github.com/angular/angular/issues/17274
- https://github.com/angular/angular/issues/5763
- https://github.com/angular/angular/issues/17092

**Comment by wizarrc** [comment]
@tytskyi 
1. I updated the typo in the title. I will change to ngClass everywhere shortly, as this is not ngStyle.
2. Thanks for the workarounds. I should have listed it along with my bug report.

I looked at the 4 related issues you listed, and all but the top one is closed. Does that mean they are already fixed or won't fix? The top related issue mentions using both ngClass and class. I am only using ngClass in my repo. I think this is a separate issue.

Anyways, here is a precise case where it fails, and is clearly a bug. If you debug the code, you can see that it adds the class to both the add and remove list during the `ngDoCheck` call in ngClass.

**Comment by vicb** [comment]
That is definitely a bug in the iterable differ

**Comment by wizarrc** [comment]
@vicb I think you are right.

https://github.com/angular/angular/blob/master/packages/core/src/change_detection/differs/default_iterable_differ.ts#L196

On the third item check, I took a screen shot before and after it calls `_verifyReinsertion`. It looks like the record it is returning is also being used to progress to the next record, which in this case, is causing the record to show it being added and then deleted.

Before:
<img width="649" alt="before-verifyreinsertion" src="https://user-images.githubusercontent.com/8039397/27719388-16f8be9e-5d20-11e7-8399-1021d78faec0.png">

After:
<img width="644" alt="after-verifyreinsertion" src="https://user-images.githubusercontent.com/8039397/27719391-1ae94a64-5d20-11e7-9f0d-d9f992ef8962.png">

Could the fix be as simple as the following?
```
// Create a temp variable so as to not affect the next record in the loop
let checkRecord = record;
if (mayBeDirty) {
    // TODO(misko): can we limit this to duplicates only?
    checkRecord = this._verifyReinsertion(checkRecord, item, itemTrackBy, index);
}
if (!looseIdentical(checkRecord.item, item)) this._addIdentityChange(checkRecord, item);
```
so record is not changed for when it calls `record = record._next;'.

**Comment by wizarrc** [comment]
I updated the repo to show duplicated data using only the `DefaultIterableDiffer` along with output from the added and removed list..

http://plnkr.co/edit/WPWy2Nx0B4LErW7dxQKK?p=preview

**Comment by vicb** [comment]
Thanks guys for helping in debugging this issue, that's really appreciated !

@wizarrc the fix is probably along the lines of what you have proposed. I'll double check tomorrow.

**Comment by wizarrc** [comment]
@vicb No problem. Once I figure out the build system, I could test my proposals and submit PRs, until then, that's the best I can do.

**Comment by vicb** [comment]
Your are probably looking for `./test.sh browserNoRouter` ?

**Comment by wizarrc** [comment]
@vicb I'm pretty sure that's bash talk. I'm coming from the command prompt and PowerShell (bat/ps1 scripts) world. I don't suppose there is vanilla Windows support for building angular? Once Bash for Windows is enabled by default, I will be taking a look into that, but right now it's still kinda beta.

**Comment by wizarrc** [comment]
Getting off topic, but I'd like to mention that in my experience so far in build systems for the web include systemjs or webpack. It appears that angular has an entirely unique build system, that requires shell scripting, or am I missing something? Other projects I can run "npm run build" "npm run test" for example.

**Comment by vicb** [comment]
we all agree that the current build system is far from ideal. 
the team is working on introducing bazel.

**Comment by wizarrc** [comment]
@vicb I tried using the bash shell that comes installed as part of visual studio, part of git for windows. I got some weird errors. Is there any prep instructions for using Windows? I tried npm install first, then `./build.sh` from the bash client giving me a 'failed to create symbolic link' error. Is this script related to that error?
https://github.com/angular/angular/blob/master/scripts/windows/create-symlinks.sh

Afterwards I typed the command `./test.sh browserNoRouter` you supplied, also gave errors.

<img width="454" alt="bash-angular-build-test" src="https://user-images.githubusercontent.com/8039397/27752454-16fa2ae2-5daf-11e7-8534-f7d518d8a982.png">

**Comment by vicb** [comment]
I'll have a PR ready in 10mn, could you please take a look and tell me what you think ?

**Comment by wizarrc** [comment]
@vicb LGTM

I think I understand better how the differ works now. It's a complicated data structure. I'd have to really break it down sometime when I need a good puzzle.

**Comment by vicb** [comment]
Thanks for the feedback - and contact me if you like puzzles we probably can find some issues for you to work on :)

**Comment by wizarrc** [comment]
@vicb I'd love to, right now I am still having lots of issues getting angular to build on windows. The `build.sh` script added `rsync` commands back in February and I think I found workarounds, including manually adding a junction directory that's out of date in the `scripts\windows\create-symlinks.sh` script, upgrading node to version 6, even though the `DEVELOPER.md` says version < 6, and modifying the build script to workaround using root directories with `rsync`. I'm still hanging the process with warning messages. Windows support seems to be a real mess at the moment.

I posted my progress here: https://github.com/angular/angular/issues/17894#issuecomment-312512265

Any help getting that working would greatly help me to get involved.
